examples/scatterplot-renderer.js

```diff
@@ -0,0 +1,148 @@
+  size: dotSize = 'population',
+  color: colorProperty = 'region'
+} = {}) => {
+  const svgRenderer = createSvgRenderer({ width, height });
+
+  const createXPos = domain =>
+    d3
+      .scaleLinear()
+      .domain(domain)
+      .rangeRound([DEFAULT_PADDING / 2, DEFAULT_SIZE - DEFAULT_PADDING / 2]);
+
+  const createYPos = domain =>
+    d3
+      .scaleLinear()
+      .domain(domain)
+      .rangeRound([DEFAULT_SIZE - DEFAULT_PADDING / 2, DEFAULT_PADDING / 2]);
+
+  const createZPos = domain =>
+    d3
+      .scaleLinear()
+      .domain(domain)
+      .range(DEFAULT_DOTSIZE_RANGE);
+
+  const createColorScale = domain =>
+    d3
+      .scaleOrdinal()
+      .domain(domain)
+      .range(DEFAULT_COLOR_RANGE);
+
+  const createScatterplot = (xDomain, yDomain, sizeDomain, color) => {
+    const svg = d3
+      .create('svg')
+      .attr('viewBox', `0 0 ${DEFAULT_SIZE} ${DEFAULT_SIZE}`);
+
+    const xPos = createXPos([Math.min(...xDomain), Math.max(...xDomain)]);
+
+    const xAxis = svg
+      .append('g')
+      .attr('transform', `translate(0, ${DEFAULT_PADDING / 2})`);
+
+    xAxis
+      .call(
+        d3
+          .axisBottom(xPos)
+          .ticks(3)
+          .tickSize(DEFAULT_SIZE - DEFAULT_PADDING)
+      )
+      .call(g => g.select('.domain').remove())
+      .call(g => g.selectAll('.tick line').attr('stroke', DEFAULT_LINE_COLOR))
+      .call(g =>
+        g
+          .selectAll('.tick text')
+          .attr('fill', DEFAULT_TEXT_COLOR)
+          .attr('font-size', 'smaller')
+      );
+
+    const yPos = createYPos([Math.min(...yDomain), Math.max(...yDomain)]);
+
+    const yAxis = svg
+      .append('g')
+      .attr('transform', `translate(${DEFAULT_PADDING / 2}, 0)`);
+
+    yAxis
+      .call(
+        d3
+          .axisLeft(yPos)
+          .ticks(3)
+          .tickSize(DEFAULT_PADDING - DEFAULT_SIZE)
+      )
+      .call(g => g.select('.domain').remove())
+      .call(g => g.selectAll('.tick line').attr('stroke', DEFAULT_LINE_COLOR))
+      .call(g =>
+        g
+          .selectAll('.tick text')
+          .attr('fill', DEFAULT_TEXT_COLOR)
+          .attr('font-size', 'smaller')
+          .attr('writing-mode', 'vertical-lr')
+          .attr('x', -6)
+          .attr('dy', 12)
+      );
+
+    const cell = svg.append('g');
+
+    cell
+      .append('rect')
+      .attr('fill', 'none')
+      .attr('stroke', DEFAULT_TEXT_COLOR)
+      .attr('x', DEFAULT_PADDING / 2 + 0.5)
+      .attr('y', DEFAULT_PADDING / 2 + 0.5)
+      .attr('width', DEFAULT_SIZE - DEFAULT_PADDING)
+      .attr('height', DEFAULT_SIZE - DEFAULT_PADDING);
+
+    const zPos = createZPos([Math.min(...sizeDomain), Math.max(...sizeDomain)]);
+
+    const circle = cell
+      .selectAll('circle')
+      .data(xDomain)
+      .join('circle')
+      .attr('cx', d => xPos(d))
+      .data(yDomain)
+      .join('circle')
+      .attr('cy', d => yPos(d))
+      .data(sizeDomain)
+      .join('circle')
+      .attr('r', d => zPos(d));
```

Comment:
This works but it's not very D3-like. Typically what you do is bind the data to a node and use it for subsequently.

```javascript
const circle = cell.selectAll('circle').data(data).join('circle')
  .attr('cx', d => x(d[xProp]))
  .attr('cy', d => y(d[yProp]))
  .attr('r', d => r(d[radiusProp]))
  .attr('fill', d => colorMap(d[colorProp]));
```

Why do we want to do this? Well next time we call this piece of code it'll update depending on the data. E.g., when we aggregate data.